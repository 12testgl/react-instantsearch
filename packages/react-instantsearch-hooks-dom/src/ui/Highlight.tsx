import React, { Fragment } from 'react';

import { cx } from './lib/cx';

type HighlightPartProps = {
  children: React.ReactNode;
  baseClassName: string;
  highlightedTagName: React.ReactType;
  nonHighlightedTagName: React.ReactType;
  isHighlighted: boolean;
};

function HighlightPart({
  baseClassName,
  children,
  highlightedTagName,
  isHighlighted,
  nonHighlightedTagName,
}: HighlightPartProps) {
  const TagName = isHighlighted ? highlightedTagName : nonHighlightedTagName;

  return (
    <TagName
      className={
        isHighlighted
          ? `${baseClassName}-highlighted`
          : `${baseClassName}-nonHighlighted`
      }
    >
      {children}
    </TagName>
  );
}

type HighlightedPart = {
  isHighlighted: boolean;
  value: string;
};

export type HighlightProps = React.HTMLAttributes<HTMLSpanElement> & {
  baseClassName: string;
  highlightedTagName?: React.ReactType;
  nonHighlightedTagName?: React.ReactType;
  separator?: React.ReactNode;
  parts: HighlightedPart[][];
};

export function Highlight({
  baseClassName,
  parts,
  highlightedTagName = 'mark',
  nonHighlightedTagName = 'span',
  separator = ', ',
  ...props
}: HighlightProps) {
  return (
    <span {...props} className={cx(baseClassName, props.className)}>
      {parts.map((part, partIndex) => {
        const isLastPart = partIndex === parts.length - 1;

        return (
          <Fragment key={partIndex}>
            {part.map((subPart, subPartIndex) => (
              <HighlightPart
                key={subPartIndex}
                baseClassName={baseClassName}
                highlightedTagName={highlightedTagName}
                nonHighlightedTagName={nonHighlightedTagName}
                isHighlighted={subPart.isHighlighted}
              >
                {subPart.value}
              </HighlightPart>
            ))}

            {!isLastPart && (
              <span className={`${baseClassName}-separator`}>{separator}</span>
            )}
          </Fragment>
        );
      })}
    </span>
  );
}
