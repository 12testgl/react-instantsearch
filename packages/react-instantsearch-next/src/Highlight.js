import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Highlight extends Component {
  static propTypes = {
    hit: PropTypes.shape({
      _highlightResult: PropTypes.object.isRequired,
    }).isRequired,
    attribute: PropTypes.string.isRequired,
  };

  replaceHighlightTags() {
    const { attribute, hit } = this.props;

    // It doesn't work at all this methods but that's okay for the POC
    return hit._highlightResult[attribute].value
      .replace('<em>', '<span class="ais-Highlight-highlighted">')
      .replace('</em>', '</span>');
  }

  render() {
    return (
      <span
        className="ais-Highlight"
        dangerouslySetInnerHTML={{ __html: this.replaceHighlightTags() }}
      />
    );
  }
}

export default Highlight;
