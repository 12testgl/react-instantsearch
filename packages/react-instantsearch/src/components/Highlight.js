import PropTypes from 'prop-types';
import React from 'react';
import Highlighter from './Highlighter';

export default function Highlight(props) {
  return (
    <Highlighter
      highlightProperty="_highlightResult"
      widgetClassName="Highlight"
      {...props}
    />
  );
}

Highlight.propTypes = {
  hit: PropTypes.object.isRequired,
  attributeName: PropTypes.string.isRequired,
  highlight: PropTypes.func.isRequired,
  tagName: PropTypes.string,
};
