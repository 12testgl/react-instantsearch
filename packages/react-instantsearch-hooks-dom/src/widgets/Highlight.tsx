import {
  getHighlightedParts,
  getPropertyByPath,
} from 'instantsearch.js/es/lib/utils';
import React from 'react';

import { Highlight as HighlightUiComponent } from '../ui/Highlight';

import type { PartialKeys, ClassNames } from '../types';
import type { HighlightProps as HighlightUiComponentProps } from '../ui/Highlight';
import type { BaseHit, Hit } from 'instantsearch.js';

export type HighlightProps<THit extends Hit<BaseHit>> = {
  hit: THit;
  attribute: keyof THit | string[];
} & PartialKeys<
  Omit<HighlightUiComponentProps, 'parts' | 'classNames'>,
  'highlightedTagName' | 'nonHighlightedTagName' | 'separator'
> &
  ClassNames<HighlightUiComponentProps>;

export function Highlight<THit extends Hit<BaseHit>>({
  hit,
  attribute,
  classNames = {},
  highlightedTagName,
  nonHighlightedTagName,
  separator,
  ...props
}: HighlightProps<THit>) {
  const property =
    getPropertyByPath(hit._highlightResult, attribute as string) || [];
  const properties = Array.isArray(property) ? property : [property];

  const parts = properties.map((singleValue) =>
    getHighlightedParts(singleValue.value || '')
  );

  return (
    <HighlightUiComponent
      {...props}
      classNames={{
        root: ['ais-Highlight', classNames.root],
        highlighted: ['ais-Highlight-highlighted', classNames.highlighted],
        nonHighlighted: [
          'ais-Highlight-nonHighlighted',
          classNames.nonHighlighted,
        ],
        separator: ['ais-Highlight-separator', classNames.separator],
      }}
      parts={parts}
      highlightedTagName={highlightedTagName}
      nonHighlightedTagName={nonHighlightedTagName}
      separator={separator}
    />
  );
}
