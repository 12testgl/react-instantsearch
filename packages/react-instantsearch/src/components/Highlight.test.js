import React from 'react';
import renderer from 'react-test-renderer';
import parseAlgoliaHit from '../core/highlight';
import Highlight from './Highlight';

describe('Highlight', () => {
  it('parses an highlighted attribute of hit object', () => {
    const hitFromAPI = {
      objectID: 0,
      deep: { attribute: { value: 'awesome highlighted hit!' } },
      _highlightResult: {
        deep: {
          attribute: {
            value: {
              value:
                'awesome <ais-highlight>hi</ais-highlight>ghlighted <ais-highlight>hi</ais-highlight>t!',
              fullyHighlighted: true,
              matchLevel: 'full',
              matchedWords: [''],
            },
          },
        },
      },
    };

    const highlight = ({ hit, attributeName, highlightProperty }) =>
      parseAlgoliaHit({
        preTag: '<ais-highlight>',
        postTag: '</ais-highlight>',
        attributeName,
        hit,
        highlightProperty,
      });

    const tree = renderer.create(
      <Highlight
        cx={(...x) => x.join(' ')}
        attributeName="deep.attribute.value"
        hit={hitFromAPI}
        highlight={highlight}
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders a hit with a custom tag correctly', () => {
    const hitFromAPI = {
      objectID: 0,
      deep: { attribute: { value: 'awesome highlighted hit!' } },
      _highlightResult: {
        deep: {
          attribute: {
            value: {
              value:
                'awesome <ais-highlight>hi</ais-highlight>ghlighted <ais-highlight>hi</ais-highlight>t!',
              fullyHighlighted: true,
              matchLevel: 'full',
              matchedWords: [''],
            },
          },
        },
      },
    };

    const highlight = ({ hit, attributeName, highlightProperty }) =>
      parseAlgoliaHit({
        preTag: '<ais-highlight>',
        postTag: '</ais-highlight>',
        attributeName,
        hit,
        highlightProperty,
      });

    const tree = renderer.create(
      <Highlight
        cx={(...x) => x.join(' ')}
        attributeName="deep.attribute.value"
        hit={hitFromAPI}
        highlight={highlight}
        tagName="marquee"
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders a hit with a custom className', () => {
    const hitFromAPI = {
      objectID: 0,
      deep: { attribute: { value: 'awesome highlighted hit!' } },
      _highlightResult: {
        deep: {
          attribute: {
            value: {
              value:
                'awesome <ais-highlight>hi</ais-highlight>ghlighted <ais-highlight>hi</ais-highlight>t!',
              fullyHighlighted: true,
              matchLevel: 'full',
              matchedWords: [''],
            },
          },
        },
      },
    };

    const highlight = ({ hit, attributeName, highlightProperty }) =>
      parseAlgoliaHit({
        preTag: '<ais-highlight>',
        postTag: '</ais-highlight>',
        attributeName,
        hit,
        highlightProperty,
      });

    const tree = renderer.create(
      <Highlight
        cx={(...x) => x.join(' ')}
        className="MyCustomHighlight"
        attributeName="deep.attribute.value"
        hit={hitFromAPI}
        highlight={highlight}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
