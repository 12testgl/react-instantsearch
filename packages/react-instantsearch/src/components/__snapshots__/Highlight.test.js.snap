// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlight parses an highlighted attribute of hit object 1`] = `
<span
  className=""
>
  <span
    className="nonHighlighted"
  >
    awesome 
  </span>
  <em
    className="highlighted"
  >
    hi
  </em>
  <span
    className="nonHighlighted"
  >
    ghlighted 
  </span>
  <em
    className="highlighted"
  >
    hi
  </em>
  <span
    className="nonHighlighted"
  >
    t!
  </span>
</span>
`;

exports[`Highlight renders a hit with a custom tag correctly 1`] = `
<span
  className=""
>
  <span
    className="nonHighlighted"
  >
    awesome 
  </span>
  <marquee
    className="highlighted"
  >
    hi
  </marquee>
  <span
    className="nonHighlighted"
  >
    ghlighted 
  </span>
  <marquee
    className="highlighted"
  >
    hi
  </marquee>
  <span
    className="nonHighlighted"
  >
    t!
  </span>
</span>
`;
